<a href="https://colab.research.google.com/github/jgbell100/P5-stars/blob/master/lesson4_data_manipulation_of_columns.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Manipulation - Manipulating Columns in a Dataset


This is a chance for you to try manipulating datasets columns using Python. We'll be using the same datasets that you saw in the slides, so you can always refer back to there too.

In these exercises you will practice:

* *selecting* the columns you want, and
* *reordering* the columns  

## Set up

To get set up, load the Python packages and datasets you'll be using into this notebook. Run the code cells in the **Load the Python packages that are needed** and **Upload the datasets** sections below to do this.

### Load the Python packages that are needed

In [30]:
# pandas is the most important package for data manipulation
import pandas as pd

# these packages will be used to upload datasets
from os import path
from google.colab import files

In [31]:
input_file_path = "./"

### Upload the datasets

Upload the following datasets from [your computer]:

* `lochs.csv` (the volume, length and maximum depth of some Scottish lochs)
* `cities.csv` (the temperature and time of day when dawn and dusk is, for some world-wide cities) 
* `clothes.csv` (the type, size, price and colour of clothes)
* `used_cars.csv` (the make, model, year of registration, price and approval rating of some used cars)

by running the code cell below.

You'll be prompted to chose the files you want to upload. Choose `lochs.csv` `cities.csv`, `clothes.csv` and `used_cars.csv`.

____________________
You can select all of the files to upload at once.

On a Mac: *hold down the command key and then click all the files you want.*

On Windows: *click the first file, hold down the Ctrl key and then click all the files you want.*
____________________



In [3]:
# [This is a colab-specific method for uploading datasets.]
# [A better method needs to be found to enable learners to load the datasets.]

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving used_cars.csv to used_cars.csv
Saving billionaires.csv to billionaires.csv
Saving cities.csv to cities.csv
Saving clothes.csv to clothes.csv
Saving dawn.csv to dawn.csv
Saving lochs.csv to lochs.csv
Saving lochs2.csv to lochs2.csv
Saving mountains.csv to mountains.csv
Saving weather_forecast.csv to weather_forecast.csv
User uploaded file "used_cars.csv" with length 160 bytes
User uploaded file "billionaires.csv" with length 318140 bytes
User uploaded file "cities.csv" with length 122 bytes
User uploaded file "clothes.csv" with length 92 bytes
User uploaded file "dawn.csv" with length 61 bytes
User uploaded file "lochs.csv" with length 160 bytes
User uploaded file "lochs2.csv" with length 127 bytes
User uploaded file "mountains.csv" with length 2373 bytes
User uploaded file "weather_forecast.csv" with length 144 bytes


Read each of the datasets from csv files into a pandas dataframe.

In [4]:
lochs = pd.read_csv(path.join(input_file_path, "lochs.csv"))
cities = pd.read_csv(path.join(input_file_path, "cities.csv"))
clothes = pd.read_csv(path.join(input_file_path, "clothes.csv"))
used_cars = pd.read_csv(path.join(input_file_path, "used_cars.csv"))

Have a look at each of the datasets to make sure they contain what you expect.

In [5]:
lochs

loch   volume  length  max_depth
0    Loch Ness  7.45280    36.2        227
1  Loch Lomond  2.61826    36.0        190
2   Loch Morar  2.38560    18.8        310
3     Loch Tay  1.67854    23.0        150
4     Loch Awe  1.20076    41.0         94

In [6]:
cities

temperature   dusk  dawn   location
0           15  22:50  3:30  Edinburgh
1           21  22:28  5:09      Paris
2           17  17:21  6:25     Sydney
3           17  20:55  4:54   New York

In [7]:
clothes

type size  price_£ colour
0   Hat    M    15.99    Red
1  Coat    L    44.95   Blue
2   Hat    S    10.99  Black
3   Top    S    20.00  White

In [8]:
used_cars

make    model  price  rating
0      Audi       A1  12500      62
1      Ford   Fiesta  12000      48
2      Ford     Puma  22500      71
3  Vauxhall    Corsa  11995      74
4    Toyota  Corolla  22075      84
5  Mercedes  S Class  49948      81

If everything looks okay, we're ready to do some data manipulation!

(If not, try uploading the csv files again).

## Manipluating Columns

### Select columns 

You're doing some analysis of Scottish lochs. You're only interested in the name of the `loch` and its `length`.

From the `lochs` dataset, select the `loch` and `length` columns and make a new dataframe, `lochs_small` from these columns.

Display this new dataframe (so that you can see what it looks like).

#### Example

In [9]:
#@title
# The next line does the select
lochs_small = lochs[['loch', 'length']]

# The next line displays the new dataframe 
lochs_small

loch  length
0    Loch Ness    36.2
1  Loch Lomond    36.0
2   Loch Morar    18.8
3     Loch Tay    23.0
4     Loch Awe    41.0

#### Task 1

1. Select the name of the `loch`, its `volume` and its maximum depth (`max_depth`) and save it to a new dataframe.

2. Display the new dataframe.



In [10]:
# select the columns


# display the new dataframe 


#### Extension Task

1. Choose any columns from the `cities` dataframe.
2. Select them and save the new columns to a new dataframe
2. Display the new dataframe.


In [11]:
# select the columns


# display the new dataframe 


### Reorder columns

You've found a dataset on cities which is a little hard to read because of the way the columns are ordered. You decide to re-order them, to make the data easier to read.

#### Example

Reorder the columns in the `cities` dataset so that they are in the order, `location`, `temperature`, `dawn`, `dusk`. 

Save the new dataframe that has been created by reordering.

Display the new dataframe.

In [12]:
# reorder the columns
cities_new = cities[['location', 'temperature', 'dawn', 'dusk']]

# display the new dataframe
cities_new

location  temperature  dawn   dusk
0  Edinburgh           15  3:30  22:50
1      Paris           21  5:09  22:28
2     Sydney           17  6:25  17:21
3   New York           17  4:54  20:55

That's better!

(Did you notice that *selecting* and *reordering* are really similar?)

#### Task 1

Another sensible way to reorder the columns so that they're easier to understand would be `location`, `dawn`, `dusk`, `temperature`.  

1. Reorder the columns in the `cities` dataset so that they are in this order 
2. Save the new dataframe that has been created by reordering.
3. Display the new dataframe.

In [13]:
# reorder the columns


# display the new dataframe 


#### Extension Task

1. Reorder the columns from the `lochs` dataframe.
2. Save the columns to a new dataframe.
2. Display the new dataframe.


In [14]:
# reorder the columns


# display the new dataframe 


## Reformat Columns

Sometimes we may wish to change how the values in a column in data are *displayed* without changing the *actual* values. 

#### Example 1 

Take the `price_£` column in the `clothes` dataframe, for example. Run the following code cell to see how the values in this column are displayed.

In [15]:
clothes

type size  price_£ colour
0   Hat    M    15.99    Red
1  Coat    L    44.95   Blue
2   Hat    S    10.99  Black
3   Top    S    20.00  White

<a name="format-dictionary"></a>
You can see that they are displayed as floating point numbers. What if we wanted to display them with a '£' symbol e.g. **£15.99** rather than **15.99**? How could we do that?

We can use the pandas `format()` function. We pass a Python dictionary to `format()`. The dictionary tells `format()` how to format the columns you wish to format. It has column names as keys, and the format to use as values. This lets us specify how we'd like each column to be formatted.

In the examples below, we will keep it simple and only format a single column at a time.

Here's an example of a dictionary we can use to display values in the `price_£` column so that they start with a '£' and are shown with 2 decimal places.

```python
pound_to_2_dp_formatter = {
    'price_£': '£{:,.2f}'
}
```

The format '£{:,.2f}' is a string. 

Run the next code cell to create the format dictionary.

In [16]:

# This is the dictionary
pound_to_2_dp_formatter = {
    # price_£ is the column we want to format
    # and £{:,.2f} is the format we want to use for it.
    # This format means "Put a £ symbol at the start and display 2 decimal places after the decimal point."
    'price_£': '£{:,.2f}'
}

Now we can use this dictionary to change how the values in the `price_£` column are displayed. 

Run the next code cell to see.

In [17]:
# format the dataframe 'clothes' using the format dictionary you've created
clothes.style.format(pound_to_2_dp_formatter)

Let's make sure that it's only how the column `price_£` is *displayed* that has changed, and not the *actual* values.

Run the next code cell to check.

In [18]:
clothes

type size  price_£ colour
0   Hat    M    15.99    Red
1  Coat    L    44.95   Blue
2   Hat    S    10.99  Black
3   Top    S    20.00  White

Are the prices still floating point numbers with no '£' symbol? Great!

#### Task 1

In the `used_cars` dataset, each car has an `approval_rating`. This is the percentage of customers who bought this car who would recommend it to others. 

Run the next code cell to have a look.







In [22]:
used_cars

make    model  price  rating
0      Audi       A1  12500      62
1      Ford   Fiesta  12000      48
2      Ford     Puma  22500      71
3  Vauxhall    Corsa  11995      74
4    Toyota  Corolla  22075      84
5  Mercedes  S Class  49948      81

However, when you look at the dataframe, it's not obvious that the values in the `approval_rating` column are percentages, is it? 

Make this clearer by displaying these values with a '%' symbol e.g. 62%.

The format string you will need to do this is `'{:,.0f}%'` (remember to include the single quotes).

**Step 1**

Create a format dictionary like the one we created [here](#format-dictionary).

Call it `percentage_formatter`.


In [23]:
# write your code here to create the format dictionary here


##### Hint

In [24]:
#@title
# What is the name of the column you want to format?

**Step 2**

Use the format dictionary to format the column for the approval rating.

##### Hint 1

In [25]:
#@title
# your format dictionary should look like this

percentage_formatter = {
    'approval_rating': '{:,.0f}%'
}

##### Hint 2

In [26]:
#@title
# pass your format dictionary to the used_cars.style.format() dictionary

In [27]:
# write your code here to format the column for the approval rating with a % symbol.


#### Task 2

Try reformatting the values in a column yourself. In the `used_cars` dataset, display the price of the car in £.

The format dictionary that you will need to create is very similar to one we used for the `clothes` dataframe, described [here](#format-dictionary).

The format string that you will need is the same as the one we used above to reformat the `price_£` column in the `clothes` dataframe: `'£{:,.2f}'` (remember to include the single quotes).

If you need some help have a look at the hints below.

##### Hints

In [ ]:
#@title
# which column do you need to reformat?


In [34]:
#@title
# create a new format dictionary that is similar to 
# the pound_to_2_dp_formatter that we created above. 
# call it used_cars_pound_to_2_dp_formatter


In [ ]:
#@title
# Use the format dictionary used_cars_pound_to_2_dp_formatter to format the column for the price.

In [21]:
# Write your code here


#### Extension Task

In the `lochs` dataframe the `volume` column holds the volume of water in each of the lochs (in $m^{3}$).

Run the next cell to have a look. 

In [28]:
lochs

loch   volume  length  max_depth
0    Loch Ness  7.45280    36.2        227
1  Loch Lomond  2.61826    36.0        190
2   Loch Morar  2.38560    18.8        310
3     Loch Tay  1.67854    23.0        150
4     Loch Awe  1.20076    41.0         94

As you can see, the `volume` measurements are accurate to 5 decimal places. That's very accurate, and great for scientific research, but you'd like to share your results with the members of a stand up paddle board club, who might just get a bit confused by such large numbers. 

You decide to display the volume of each loch as accurate to 1 decimal place. 

The format string you will need for this is `'{:.1f}'` (remember to include the single quotes). 

In [29]:
# write your code here


(Did you notice that the `format()` function also takes care of *rounding* too?)